### PR TITLE
GitHub Actions: Copy draft release notes to CHANGELOG

### PR DESCRIPTION
**Pull request by greg-finley**
### Brief summary of the change made

Following https://github.com/sqlfluff/sqlfluff/pull/3347

It finds the latest draft release in the repo (which should be the right one) and puts the markdown into the CHANGELOG. Example here: https://github.com/greg-finley/sqlfluff/pull/25

Latest CI run: https://github.com/greg-finley/sqlfluff/runs/6484054427?check_suite_focus=true

Makes progress on https://github.com/sqlfluff/sqlfluff/issues/1422

### Are there any other side effects of this change that we should be aware of?
It does not have "New Contributors" (separate PR / punt on it).

It still needs the part of fixing all the links (let's do as a separate PR).

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: .github/workflows/create-release-pull-request.yaml [hunk 1]
```diff
@@ -21,10 +21,18 @@ jobs:
...
+          echo ${{ secrets.GITHUB_TOKEN }} | gh auth login --with-token
+          RELEASE_LIST=$(gh release list)
+          echo $RELEASE_LIST
+          LATEST_DRAFT_RELEASE=$(echo $RELEASE_LIST | grep Draft | head -n1 | awk '{print $1;}')
+          echo $LATEST_DRAFT_RELEASE
+          echo "" > notes.txt
+          echo "##[${{ github.event.inputs.newVersionNumber }}] - $(date +%F)" >> notes.txt
+          echo "" >> notes.txt
+          gh release view $LATEST_DRAFT_RELEASE | sed '1,/^--$/d' >> notes.txt
+          cat notes.txt
```
Comment (greg-finley) [inline .github/workflows/create-release-pull-request.yaml:33]: I think it's ok to leave this stuff printing logs to the GitHub Actions machine, but we can remove if you prefer.

File: .github/workflows/create-release-pull-request.yaml [hunk 2]
```diff
@@ -21,10 +21,18 @@ jobs:
 
       - name: Start Changelog entry
         run: |
-          function dt {
-            date +'%Y-%m-%d'
-          }
-          sed -i "s/<!--Start Of Releases (DO NOT DELETE THIS LINE)-->/<!--Start Of Releases (DO NOT DELETE THIS LINE)-->\n\n## [${{ github.event.inputs.newVersionNumber }}] - $(date +%F)\n\n## Highlights\n\n## What’s Changed\n\n## New Contributors/" ./CHANGELOG.md
+          echo ${{ secrets.GITHUB_TOKEN }} | gh auth login --with-token
```
Comment (tunetheweb) [inline .github/workflows/create-release-pull-request.yaml:24]: `echo`ing secrets always scares me (thought GitHub is pretty good at masking them).

This https://docs.github.com/en/actions/using-workflows/using-github-cli-in-workflows seems to suggest you can use syntax like this instead:

```
jobs:
  comment:
    runs-on: ubuntu-latest
    steps:
      - run: gh issue comment $ISSUE --body "Thank you for opening this issue!"
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          ISSUE: ${{ github.event.issue.html_url }}
```

**Review by tunetheweb — CHANGES_REQUESTED** [review]

File: .github/workflows/create-release-pull-request.yaml (hunk 1 above)
Comment (tunetheweb) [reply .github/workflows/create-release-pull-request.yaml:33]: Yeah that's fine. As long as it doesn't contain sensitive data (which is doesn't here).

File: .github/workflows/create-release-pull-request.yaml (hunk 2 above)
Comment (greg-finley) [reply .github/workflows/create-release-pull-request.yaml:24]: FWIW, it did mask, but yeah, I will try this out.

Screen Shot 2022-05-17 at 11 48 39 PM

**Comment by tunetheweb** [comment]
BTW one thing to consider, would be that in future it would be great to be able to update the branch and replace the list with the latest list.

E.g.

- Open PR automatically with current list of draft release
- Merge some more PRs
- Somehow auto update the release PR with latest list of changes (or by completely replacing the current list with the latest draft).

So keep that in mind if we want a `<!-- Start Latest PRs -->` and `<!-- End Latest Merged PRs -->` lines to easily identify that. Or maybe can just search for first `What's Changed` and delete until newline?

File: .github/workflows/create-release-pull-request.yaml (hunk 2 above)
Comment (greg-finley) [reply .github/workflows/create-release-pull-request.yaml]: OK, this worked on my end: https://github.com/sqlfluff/sqlfluff/pull/3360/commits/bdca2a05a25a5dd37b1b06483a4e2093199b813f

File: .github/workflows/create-release-pull-request.yaml [hunk 3]
```diff
@@ -21,10 +21,19 @@ jobs:
 
       - name: Start Changelog entry
         run: |
-          function dt {
-            date +'%Y-%m-%d'
-          }
-          sed -i "s/<!--Start Of Releases (DO NOT DELETE THIS LINE)-->/<!--Start Of Releases (DO NOT DELETE THIS LINE)-->\n\n## [${{ github.event.inputs.newVersionNumber }}] - $(date +%F)\n\n## Highlights\n\n## What’s Changed\n\n## New Contributors/" ./CHANGELOG.md
```
Comment (tunetheweb) [inline .github/workflows/create-release-pull-request.yaml:27]: Could we still add a blank `Highlights` and `New Contributors` section still? Seem to have lost these with this PR.

**Comment by greg-finley** [comment]
> BTW one thing to consider, would be that in future it would be great to be able to update the branch and replace the list with the latest list.
> 
> E.g.
> 
> * Open PR automatically with current list of draft release
> * Merge some more PRs
> * Somehow auto update the release PR with latest list of changes (or by completely replacing the current list with the latest draft).
> 
> So keep that in mind if we want a `<!-- Start Latest PRs -->` and `<!-- End Latest Merged PRs -->` lines to easily identify that. Or maybe can just search for first `What's Changed` and delete until newline?

Yeah I'm thinking maybe `peter-evans/create-pull-request@v4` `branch-suffix` could be adjusted/removed so that it updates the PR instead of opening new ones.

Hopefully my PRs aren't too small (getting CI stuff right is hard!) but I'd vote to punt on this idea for now and to break the original issue into several smaller ones, such as: 1) GitHub Action updates the existing PR, 2) parse the markdown into links, 3) New Contributors section, etc. I think each incremental piece we get done at least helps a little.

**Comment by tunetheweb** [comment]
Agreed. If you could add those Highlights and New Contributors sections back in then think this is good to merge.

File: .github/workflows/create-release-pull-request.yaml (hunk 3 above)
Comment (greg-finley) [reply .github/workflows/create-release-pull-request.yaml:27]: https://github.com/sqlfluff/sqlfluff/pull/3360/commits/5df1f207949b8d9f7ac0b0996ec60ebbef5cb5d6

Resulted in: https://github.com/greg-finley/sqlfluff/pull/28

**Review by tunetheweb — APPROVED** [review]
LGTM!

**Pull request merged**